### PR TITLE
fix problem with unset containers

### PR DESCRIPTION
**Pull request by mccanne**
The zeek parser was parsing unset and empty sets to the same zson.
We fixed this and realized the zson text format did not have
a way to represent an unset container.  So we added "*" to the
syntax to mean this like "-" means unset value for scalars.

Updated the parsing code to handle "*" and also updated
zson over json to encode unset containers as JSON empty object
inside of the JSON value arrays.  With the zval builder properly
handling empty container as empty container instead unset container,
a bug in ndjson was exposed where calling marshal on sets or vectors
didn't work right because the zeek value had its arrays unitialized.

Finally, updated zson spec.

### CODE REVIEW

File: pkg/zson/zval.go [hunk 1]
```diff
@@ -51,6 +51,9 @@ func ZvalToZeekString(typ zeek.Type, val []byte, isContainer bool) string {
 	var s string
 	switch typ.(type) {
 	case *zeek.TypeSet, *zeek.TypeVector:
+		if val == nil {
+			return "-"
+		}
```
Comment (nwt) [inline pkg/zson/zval.go]: Instead of adding this, how about changing line 48 from `if val == nil && !isContainer {` to `if val == nil {`?

Comment (mccanne) [reply pkg/zson/zval.go]: yes

Comment (mccanne) [reply pkg/zson/zval.go]: yes

File: pkg/zeek/type.go [hunk 2]
```diff
@@ -180,10 +180,9 @@ func parseType(in string) (string, Type, error) {
 
 // Utilities shared by compound types (ie, set and vector)
 
-// If the passed-in type is a container, ContainedType() returns
-// the type of individual elements (and true for the second value).
-// Otherwise, returns nil, false.
-func ContainedType(typ Type) Type {
+// InnerType returns the type of its elements for set and vector types
```
Comment (nwt) [inline pkg/zeek/type.go]:
```suggestion
// InnerType returns the element type for set and vector types
```

File: pkg/zeek/type.go [hunk 3]
```diff
@@ -180,10 +180,9 @@ func parseType(in string) (string, Type, error) {
...
+// or nil if the type  is not a set or vector.
```
Comment (nwt) [inline pkg/zeek/type.go]:
```suggestion
// or nil if the type is not a set or vector.
```

File: pkg/zeek/type.go [hunk 4]
```diff
@@ -194,6 +193,22 @@ func ContainedType(typ Type) Type {
 	}
 }
 
+// ContainerType returns the inner type for set and vector types in the first
```
Comment (nwt) [inline pkg/zeek/type.go]:
```suggestion
// ContainedType returns the inner type for set and vector types in the first
```

File: pkg/zeek/type.go [hunk 5]
```diff
@@ -194,6 +193,22 @@ func ContainedType(typ Type) Type {
...
+// argument and the columns of its of type for record types in the second column.
```
Comment (nwt) [inline pkg/zeek/type.go]:
```suggestion
// return value and the columns of its of type for record types in the second return value.
```

File: pkg/zeek/type.go [hunk 6]
```diff
@@ -194,6 +193,22 @@ func ContainedType(typ Type) Type {
...
+// Nil is return for both values if the type is not a set, vector, or record.
```
Comment (nwt) [inline pkg/zeek/type.go]:
```suggestion
// ContainedType returns nil for both values if the type is not a set, vector, or record.
```

File: pkg/zsio/zson/writer.go [hunk 7]
```diff
@@ -104,8 +108,11 @@ func (w *Writer) writeEscaped(val []byte) error {
 	if len(val) == 0 {
 		return nil
 	}
-	if len(val) == 1 && val[0] == '-' {
-		return w.escape('-')
+	if len(val) == 1 {
+		switch val[0] {
+		case '-':
+			return w.escape(val[0])
+		}
```
Comment (nwt) [inline pkg/zsio/zson/writer.go]: Nit: Stick with the original code.

File: pkg/zval/encoding.go [hunk 8]
```diff
@@ -46,6 +46,12 @@ func (e Encoding) Build(b []byte) ([]byte, error) {
 			return nil, err
 		}
 		if container {
+			if v == nil {
+				b = append(b, '(')
+				b = append(b, '*')
+				b = append(b, ')')
+				continue
+			}
```
Comment (nwt) [inline pkg/zval/encoding.go]: Given that we've dropped the * convention, should this still be here?

File: pkg/zeek/type.go (hunk 2 above)
Comment (mccanne) [reply pkg/zeek/type.go]: +1

File: pkg/zeek/type.go (hunk 3 above)
Comment (mccanne) [reply pkg/zeek/type.go]: +1

File: pkg/zeek/type.go (hunk 4 above)
Comment (mccanne) [reply pkg/zeek/type.go]: +1

File: pkg/zeek/type.go (hunk 5 above)
Comment (mccanne) [reply pkg/zeek/type.go]: +1

File: pkg/zsio/zson/writer.go (hunk 7 above)
Comment (mccanne) [reply pkg/zsio/zson/writer.go]: +1 bit rot

File: pkg/zval/encoding.go (hunk 8 above)
Comment (mccanne) [reply pkg/zval/encoding.go]: Yes because we're showing the difference between container-unset and value-unset which exists in the underlying zval but would be lost if we stringified them both the same way.

Comment (mccanne) [reply pkg/zval/encoding.go]: This string format doesn't conform to any sort of spec.

Comment (nwt) [reply pkg/zval/encoding.go]: In that case, you can shorten this to `b = append(b, "(*)"...)` in the future.
